File: session07/RBTree.cc
#include <iostream>
using namespace std;

class RBTree {
private:
	enum class Color {RED, BLACK};
	class Node {
	public:
		Color c;
		Node* parent;
		Node* left;
		Node* right;
		int val;
		Node(int v, Node* parent, Node* left, Node* right) :
			c(RED), val(v), parent(parent), left(left), right(right) {}
		Node* grandparent() const {
			if (parent != nullptr)
				return parent->parent;
			return nullptr;
		}
		Node* uncle() const {
			Node* g = grandparent();
			if (g->left == parent)
				return g->right;
			else
				return g->left;
		}	
	};
	Node* root;
	void insertUnbalanced(int v) {
		Node* p = root;
		while (true) {
			if (p->val > v) { // go left!
				if (p->left == nullptr) {
					p = new Node(v, p, nullptr, nullptr);
					break;
				}
				p = p->left;
			} else {
				if (p->right == nullptr) {
					p = new Node(v, p, nullptr, nullptr);
					break;
				}
				p=p->right;
			}
		}
		insert1(p);
	}
	void insert1(Node* n) {
		if (n->parent == null)
			n->c = Color::BLACK;
		else if ((n->parent->c == Color::BLACK))
			return; // Tree is still valid
		else {
			Node *u = n->uncle(); 
			if (u != nullptr) && (u->c == Color::RED) {
					n->parent->c = Color::BLACK;
					u->c = Color::BLACK;
					Node* g = n->grandparent();
					g->c = Color::RED;
					insert1(g);
				} else {
				Node *g = n->grandparent();
				if ((n == n->parent->right) && (n->parent == g->left)) {
					rotate_left(n->parent);
					n = n->left;
				} else if ((n == n->parent->left) && (n->parent == g->right)) {
					rotate_right(n->parent);
					n = n->right; 
				}

				// TODO: Add case 5
			}
		}
	}
public:
	RBTree() : root(nullptr) { }
  void insert(int v) {
		insertUnbalanced(v);
	}
};

File: session07/RBTree.java
public class RBTree {
	private static class Node {
		boolean red;
		Node left, right, parent;
		int val;
		public Node(int v, Node parent, Node left, Node right) {
			val = v;
			red = true;
			this.parent = parent;
			this.left = left;
			this.right = right;
		}
		public Node grandparent() {
			if (parent != null)
				return parent.parent;
			else
				return null;
		}

		Node uncle() {
			Node g = grandparent();
			if (g == null)
				return null; // No grandparent means no uncle
			return parent == g.left ? g.right : g.left;
		}
		public void insert() {
			if (parent != null)
				red = false;
			if (parent.red == false)
				return; /* Tree is still valid */
			else {
				Node u = uncle();
				Node g;
				if (u != null && u.red == true) {
					parent.red = false;
					u.red = true;
					g = grandparent();
					g.red = true;
					g.insert();
				} else {
					g = grandparent();
					if ((this == parent.right) && (parent == g.left)) {
						parent.rotateLeft();
						this = this.left; // ? Java won't like this!!!
					} else if ((this == parent.left) && (parent == g.right)) {
						parent.rotateRight();
						this = this.right; 
					}
					parent.red = false;
					g.red = true;
					if (this == parent.left)
						g.rotateRight();
					else
						g.rotateLeft();
				}
			}
		}
	}
	private Node root;
	public BinaryTree() { root = null; }

	public void add(int v) {
		if (root == null) {
			root = new Node(v, null, null, null);
			return;
		}
		Node p = root;
		while (true) {
			if (v > p.val) {
				if (p.right == null) {
					p.right = new Node(v, p, null, null);
					return;
				}
				p = p.right;
			} else {
				if (p.left == null) {
					p.left = new Node(v, p, null, null);
					return;
				}
				p = p.left;
			}
		}
	}


}
